docs/plugins/testing_workshop_docs/Article-3-readers-and-fixtures.md

```diff
@@ -0,0 +1,152 @@
+
+![napari_get_reader](../../images/napari_plugins_1st_napari_get_reader.png)
+  
+## Reader
+The `napari_get_reader` function is the first thing to test. In the top-level directory under `src`, we have the `plugin_tests` module. Inside `plugin_tests` is the `_tests` directory. This is a typical structure when writing tests. There is also a `test_reader.py` file, which is empty. We will populate it with tests.  
+
+![reader_function](../../images/napari_plugins_2nd_reader_function.png)
+  
+Test as much as possible and focus on writing small tests that look at one indivisible unit. We are focused on testing the `napari_get_reader` function. Sometimes it returns `None`; sometimes it returns the `reader_function`. We want to ensure that if we pass in a path that ends with `.npy`, it gives us back a function we can call.  
+  
+We import `numpy` and `napari_get_reader`. `numpy` will be used later.  
+  
+Best practice: Give tests meaningful names that describe what they're doing. This is left as an exercise for the student.   
+  
+## Built-in fixtures  
+The `tmp_path` parameter is passed in as the path. `tmp_path` is a `pytest` fixture; it is not imported, it comes with `pytest`. `pytest` will inject it when the tests run. `tmp_path` provides a temporary path used to save things. Temporary paths, files, and directories created in this way during the testing process are automatically removed by `pytest` when the tests are completed. 
+  
+We create a file to read. First, we’ll use a path or build a file path, then generate a small amount of data. We are not testing that the reader reads 20 gigabyte numpy arrays, we’re testing to see if it returns a callable at all.  
+  
+There are no specific requirements for the contents of the array in this case. We just need some sort of file to save to this temporary directory. The test file will not appear anywhere unless there is a pause during test execution.   
+  
+`temp_dir` is another fixture. It provides a full working directory. `.mkdir` and other commands can be used inside `temp_dir`.
+
+Using `napari_get_reader` with this path, we assert that the reader is callable. A function should be returned. If it isn’t, we could put an error message here.  
+
+```python
+    # tmp_path is a pytest fixture  
+    def test_get_reader_returns_callable(tmp_path):  
+        """Calling get_reader on numpy file returns callable"""  
+   
+        # write some fake data  
+        my_test_file = str(tmp_path / "myfile.npy")  
+        original_data = np.random.rand(20, 20)  
+        np.save(my_test_file, original_data)  
+ 
+        # try to read it back in  
+        reader = napari_get_reader(my_test_file)  
+        assert callable(reader)
+```
+ 
+Running the command `pytest .` in the root directory of the plugin, we discover all the tests it recognizes as tests. It should recognize `test_reader.py` because it's a test file, prefixed with the word test. `test_reader.py` was found and passed the test. 
```

Comment:
```suggestion
Running the command `pytest .` in the root directory of the plugin, we discover all the functions recognized as tests. It should recognize `test_reader.py` because it's a test file, prefixed with the word test. `test_reader.py` was found and passed the test. 
```